lib/remap-functions/src/length.rs

```diff
@@ -0,0 +1,136 @@
+use remap::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct Length;
+
+impl Function for Length {
+    fn identifier(&self) -> &'static str {
+        "length"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            accepts: |v| matches!(v, Value::Array(_) | Value::Map(_)),
+            required: true,
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required("value")?.boxed();
+
+        Ok(Box::new(LengthFn { value }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct LengthFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for LengthFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        use Value::*;
+
+        let value = self.value.execute(state, object)?;
+
+        match value {
+            Array(v) => Ok((v.len() as i32).into()),
+            Map(v) => Ok((v.len() as i32).into()),
+            _ => Err("unsupported type".into()),
+        }
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        use value::Kind;
+
+        self.value
+            .type_def(state)
+            .fallible_unless(Kind::Array | Kind::Map)
+            .with_constraint(Kind::Integer)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use value::Kind;
+
+    test_type_def![
+        value_map_infallible {
+            expr: |_| LengthFn {
+                value: map! {"foo": "bar", "baz": 27, "baq": false}.boxed()
+            },
+            def: TypeDef {
+                fallible: false,
+                kind: Kind::Integer,
+                inner_type_def: Some(
```

Comment:
I think we should be removing the `inner_type_def` in the implementation given the return type is an integer and doesn't have an inner type. I fixed a similar issue with the `includes()` function over here: https://github.com/timberio/vector/pull/5628/files

Reply:
I originally thought the same thing when writing up the `includes` function but @JeanMertz specifically requested that that be specified. I don't quite know the inner mechanics well enough to know why but I trust that there's a good reason for it.

Reply:
I _really_ need to document how to use this (and also add some QoL improvements).

But, the inner type should only be set when the `kind` is a non-scalar (e.g. maps and arrays, a.k.a. container type).

So while the input of this function might be a container type, its output will always be a scalar, and so we should never have the `inner_type_def` set to anything other than `None`, because there is nothing "inside" the scalar to look at.

Now, what's a tad confusing, is that `kind` can actually represent more than one kind (we progressively narrow the kind). So if you have `kind = Integer | Map`, then we _do_ need to keep the `inner_type_def`, since it's still relevant for when further narrowing of that kind resolves it to a map and not the integer.

What we should do though (feel free to do so in this PR), is to update the `with_constraint` to set `inner_type_def` to `None` if `kind.is_scalar()` is true.

Reply:
@JeanMertz That indeed makes sense and is consistent with what I learned working on `includes`. I've updated with the requested change.